Create a command buffer builder state tracker object

kainino0x opened the pull request:
Dependent on #18, so this will be easier to review after that's merged and this is rebased. Defer if you want.

Later (after this lands) I plan to move more of the logic from CommandBuffer.cpp to CommandBufferStateTracker.cpp.

Kangz commented:
Overall I feel the separation of concerns between the tracker and CommandBufferBuilder isn't clear enough. Ideally all the persistent state would go in the tracker (include ```aspects```) and would be  private to the tracker.

Then code in the CommandBufferBuilder could look like these examples:

```cpp
// For stuff changing the state
if (!state->BeginRenderPass(renderpass, framebuffer)) {
  return false;
}
// stuff that doesn't change the state would look like this, example for blits
if (!state->ValidateCanCopy() ||
    !state->ValidateUseBufferAs(buffer, nxt::BufferUsageBit::TransferSrc) ||
    !state->ValidateUseTextureAs(...)) {
}
here non-state dependent validation.
```

```Validate``` state methods could be const, provided we are ok making aspects mutable for lazy computation of the bindgroups aspect.

Also the state tracker could be used as a local variable inside ```ValidateGetResult``` and the relevant data (such as transitioned resources) extracted at the end of the function.

kainino0x commented:
@Kangz I've finally finished the full refactor of this state tracking and I think it goes along the guidelines you mentioned. PTAL!

Unfortunately there are a few changes which are hard to diff. I'll try to remember them:
* Removed VALIDATION_ASPECT_RENDER_PASS - I don't think it's necessary as it's not used in any of the 'fast-path' validation.
* Refactored the slow paths of ValidateCan{Dispatch,DrawArrays,DrawElements} - should be identical, though. DrawArrays and DrawElements each get their own fast path now.

All of the CommandBufferStateTracker public methods do their own HandleErrors and return bool.

Kangz commented on src/backend/common/CommandBufferStateTracker.cpp:
```diff
@@ -0,0 +1,455 @@
+    bool CommandBufferStateTracker::ValidateCanDrawArrays() {
+        // TODO(kainino@chromium.org): Check for a current render pass
+        constexpr ValidationAspects requiredAspects =
+            1 << VALIDATION_ASPECT_RENDER_PIPELINE |
```
The VERTEX_BUFFERS aspect is missing here

Kangz commented on src/backend/common/CommandBufferStateTracker.cpp:
```diff
@@ -0,0 +1,455 @@
+            return true;
+        }
+
+        return ValidateCanDraw();
```
ValidateCanDraw is a bit confusing because it is so similar to ValidateCanDrawArrays, maybe rename it, for example with RevalidateCanDraw to mean it isn't lazy

Kangz commented on src/backend/common/CommandBufferStateTracker.cpp:
```diff
@@ -0,0 +1,455 @@
+            return false;
+        }
+
+        aspects.reset(VALIDATION_ASPECT_COMPUTE_PIPELINE);
```
note: I don't know how inheritance of bindings and vertices works at renderpass start in the backing APIs, but it seems likely that we will have to reset all aspects here. This way applications will be forced to set things again in the subpass, which will remove the need to do some state-tracking in the backends.

Kangz replied:
Sounds good, it makes things clearer and should be just as fast with LTO!

Kangz commented on src/backend/common/CommandBufferStateTracker.cpp:
```diff
@@ -0,0 +1,455 @@
+            aspects.set(VALIDATION_ASPECT_RENDER_PIPELINE);
+        }
+        aspects.reset(VALIDATION_ASPECT_BIND_GROUPS);
+        aspects.reset(VALIDATION_ASPECT_VERTEX_BUFFERS);
```
Previous code was doing it, however we shouldn't reset vertex buffers here: all backing APIs keep the vertex buffers bindings on pipeline change.

Kangz commented on src/backend/common/CommandBufferStateTracker.cpp:
```diff
@@ -0,0 +1,455 @@
+    }
+
+    bool CommandBufferStateTracker::SetIndexBuffer(BufferBase* buffer) {
+        auto usage = nxt::BufferUsageBit::Index;
```
Here and for vertex buffers, we should check we are in a pipeline.
